### PR TITLE
nlp: update megatron scripts

### PR DESCRIPTION
**Pull request by amo33**
This PR updates megatron related scripts that are needed when finetuning with SFTDataset. 

One thing to point out : gpt_dataset.py script has been excluded, since it was unrelated with gpt_sft_dataset.py. 

Any feedbacks are opened, thanks.

### CODE REVIEW

File: flatflow/nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 1]
```diff
@@ -382,15 +397,23 @@ def fwd_bwd_step(self, dataloader_iter, forward_only, first_val_step=None):
             module.config.param_sync_func = param_sync_func
 
         fwd_bwd_function = get_forward_backward_func()
-
+        num_microbatches = 0
```
Comment (9rum) [inline flatflow/nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:400]: I think these lines can be simplified to:
```python
num_microbatches = 1 if self.use_flatflow else get_num_microbatches()
micro_batch_size = 1 if self.use_flatflow else get_micro_batch_size()
```

File: flatflow/nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 2]
```diff
@@ -458,7 +481,7 @@ def inference_step(self, dataloader_iter, mode):
                 # replace the last appended loss with the outputs dict
                 self.validation_step_outputs[-1] = outputs
         else:
-            if isinstance(self.trainer.test_dataloaders, list) and len(self.trainer.test_dataloaders) > 1:
```
Comment (9rum) [inline flatflow/nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:461]: I think this line is fine enough, is there a reason for the change?

**Review by 9rum — CHANGES_REQUESTED** [review]
@amo33 Thanks for filing this PR.
In addition to the comments I left, there are a few things I would like to confirm:

Regarding ``megatron_gpt_finetuning.py``,
1. It looks like the global rank ``gpu_rank`` has been obtained using ``torch.distributed.get_rank()``. Since NeMo has hierarchical communication groups, I would like to confirm if ``torch.distributed.get_rank()`` returns the global rank when there are multiple communication groups.
2. I think the memory footprint was measured using ``torch.cuda.max_memory_allocated()`` and ``torch.cuda.max_memory_reserved()`` before and after training (i.e., ``trainer.fit(model)``). I would like to confirm whether it is sufficient to track peak memory usage during training with just this.

Regarding ``gpt_sft_dataset.py``,
1. It looks like you have written ``GPTConcatSFTDataset``, which replaces ``GPTSFTDataset`` by inheriting ``flatflow.torch.utils.data.Dataset``.
2. I think this could be implemented more elaborately. That is, change ``nemo.core.classes.Dataset``, which ``GPTSFTDataset`` inherits, to inherit ``flatflow.torch.utils.data.Dataset``. This would be named to ``flatflow.nemo.core.classes.Dataset``.
3. Change the existing ``GPTSFTDataset`` to inherit ``flatflow.nemo.core.classes.Dataset``.
4. Make ``MegatronGPTSFTModel`` to use the updated ``GPTSFTDataset`` for its ``dataset_cls`` in ``megatron_gpt_sft_model.py``, line 279.

Regarding ``megatron_gpt_sft_model.py``,
I cannot find ``flatflow.torch.utils.data.distributed.DistributedBatchSampler``, which is used for ``batch_sampler``. I would like to confirm if this has been implemented but omitted in this PR.

File: flatflow/nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 2 above)
Comment (amo33) [reply flatflow/nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:461]: I must have ran ruff format in a wrong way. 
This snippet should be reverted to original code.

**Comment by amo33** [comment]
Thanks for your PR, I should fix based on your feedback.


For some of question, 
1. `flatflow.torch.utils.data.distributed.DistributedBatchSampler` is not in this PR. 
My bad I forgot to  mention.
I was expecting myself to make PR of sampler in L#15. 

2. About megatron_gpt_finetuning.py, I also think more sophisticated peak memory tracker will be helpful if there is any support. 
3. And for Question 1 so far environment is scarce , and I expect that `torch.distirbuted.get_rank()` might not return appropriate global rank.

**Review by 9rum — APPROVED** [review]
@amo33 Thanks for your hard work.
I will make ``__init__.py`` under ``flatflow/nemo/collections/nlp/models/language_modeling``. 
Let's fix ``MegatronPretrainingBatchSampler`` in the next PR.

**Pull request merged**